[s04a] Logistic regression

colinhex opened the pull request:
Hi everyone,

I expanded on the outline of the original script and added more thorough recaps & summary.

I'm happy to make further changes on request.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+    \mathbf{X}^T\mathbf{y}=\mathbf{X}^T\mathbf{X}\mathbf{w}^*
+$$
+$$
+    \frac{\mathbf{X}^T\mathbf{y}}{\mathbf{X}^T\mathbf{X}}=\mathbf{w}^*
```
You cannot divide a  vector by a matrix. You can write an inverse, then you need to assume that inverse exists. Pseudo inverse is more appropriate here.

colinhex replied:
Changed it to an inverse, where it came up.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+to another in a random image, so you indeed have to send all the data if your friend should receive it without loss. The noise is
+a state of maximum information or *maximum entropy*.
+
+When we speak of the **cross-entropy** between two probability distributions over the same set of events we are talking about
```
May be add the equation of the cross entropy between two probability distributions

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+    \sum_{i=1}^k ln\left(\prod_{c=1}^C \mu^{y_{ic}}_{ic}\right) = \sum_{i=1}^k \sum_{c=1}^C ln(\mu^{y_{ic}}_{ic}) = \sum_{i=1}^k \sum_{c=1}^C y_{ic} ln(\mu_{ic})
+$$
+
+The term $-\sum_{c=1}^C y_{ic} ln(\mu_{ic})$ is also defined as **cross-entropy**. This means that
```
Can you show why this is cross entropy?  As cross entropy is between the two probability distribution. How is the term you mentioned related to the two probability distributions?

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+
+Another way of picturing this equation is as measuring the cross-entropy between two probability distributions
+$(y_i, (1-y_i))$ and $(\mu_i, (1-\mu_i))$, as in how close the distribution of the predictions is to
+the desired, actual distribution of the data. If we examine $\mu_i$ and $(1 - \mu_i)$ we find that they simplify considerably:
```
shouldn't this be : "desired/ actual  or desired (actual)"

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+If we further consider the function for binary cases \{-1, 1\}, then we can simply write: 
+
+$$
+    NLL(\mathbf{w}) = \sum_{i=1}^k ln(1 + e^{-y_iw^Tx_i}) \quad y_i \in \{-1, 1\}
```
Elaborate a bit here. As how do we get this equation.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+    \lim_{\hat{y} \rightarrow -\infty} ln(1 + e^{-y\hat{y}}) = \begin{cases} 0 &\quad y = -1 \\ +\infty &\quad y = 1 \end{cases}
+$$
+
+If you can still remember, we derived this loss to get rid of the least squares loss, which we
```
Instead of "If you can still remember", point to the part of the text where you derived this.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+line2 = ax.plot(y, loss_01(y, y_hat), '-r', lw=1, label='L01-Loss')
+line3 = ax.plot(y, loss(y, y_hat), '-b', lw=1, label='LL-Loss')
+ax.legend(('L2-Loss', 'L01-Loss', 'Logistics-Loss'))
+plt.setp(ax, xlabel='y', ylabel='Loss for y_hat = 1', ylim=(-0.1, 2.0), xlim=(-1, 2.5))
```
Looks like,logistic loss is also different compared  to the L_{01} loss. Can you write why this logistic loss is better  compared to L_2 loss?

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+    = \mathbf{X}^T\left(\nabla_{\mathbf{w}}\frac{e^{w^T_cx}}{\sum_{c=1}^{C}e^{w^T_cx}}\right) 
+$$
+
+Now there are some good news and bad news. The good news is, we won't have to calculate this derivative and set it to
```
Why is it a  good news that we cant set it to 0?

colinhex replied:
It specifically says it is bad news that we can't set it to zero.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+Using gradient descend we can approximate local minima or maxima of functions in any number of dimensions. We iteratively
+approach the extrema by moving along the direction of the gradient of a function. You know the derivative of a function gives us
+a value for its steepness at every point in space. In gradient (steepest) descend, we find the **direction**, you 
+guessed it, of the **steepest descend** and move a distance with a certain factor, the **step size**. 
```
Maybe a bit less conversation text, would be easy to follow.  
Can you write the relation with direction of steepest descent to that of gradient. Some introductory materials on gradient descent will have this reason.

colinhex replied:
I'm sorry can you rephrase this comment? I have a hard time understanding it.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+If we start with some initial guess and repeat this process a certain number of times, where we end up must be close
+to a **local minimum**. We can formulate this in an algorithm as follows:
+
+$\textit{for} \quad t \in \{1, ..., t_{max}\} \quad \textit{do}:$
```
Put it inside a latex algorithm environment.

colinhex replied:
There is currently no LaTeX algorithm environment package in the repository. The Jupyter-Book documentation references ([here](https://sphinx-proof.readthedocs.io/en/latest/syntax.html#algorithms)) an extra extension for sphinx to enable a algorithm environment, but an inclusion in _config.yml leads to an error:

`Extension error: Could not import extension sphinx_proof (exception: No module named 'sphinx_proof')...[...]`

vinith2 replied:
Oh ok, we will look into this later, thanks for letting me know.  I will keep this issue open for now. So I remember it later.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+
+fig = plt.figure(figsize=(7,7))
+ax = plt.axes(projection='3d')
+ax.plot_surface(xx, yy, zz, cmap=cm.coolwarm,  rcount=100, ccount=100);
```
what does the color signify?  Is it the z-axis. Maybe add a colorbar as well?

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+
+If gradient descend is used as an optimization technique for a model, then the step size can be seen, among else, as its 
+**learning rate**. If we set it too high, we might overshoot, oscillate or even diverge with our weights. If we set it
+to low, we might need a *large* number of iterations to find it.
```
Can you plot the gradient descent with different step size? Essentially showing the oscillations or diverge phenomena you  are talking about.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+form expression for our problem.
+
+$$
+\mathbf{X}^T(\nabla_{\mathbf{w}}\frac{e^{w^T_cx}}{\sum_{c=1}^{C}e^{w^T_cx}}) 
```
Should this be equated to 0? I don't understand what this term signify.

colinhex replied:
This is equivalent to the partial derivative of the previous expression.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+Think of two images, one is white noise and one is of a banana. Which of these pictures carries more information? You might think
+the banana, since we recognize it, and it conjures up relations in our minds. White noise on the other hand is meaningless to us.
+
+![alt](./images/white-noise.png) ![alt](./images/Banana-Single-2.jpg)
```
Can you center this image?

colinhex replied:
Sadly, to date I have not found a way to center this. Markdown Tables and markdown resizing will fail and HTML embedding as well.

vinith2 replied:
oh ok, the current setting should be good then.

vinith2 reviewed (commented):
Hi, the page looks good. I have added some for some specific lines in the script.

colinhex commented on logistic-regression.md:
```diff
@@ -0,0 +1,752 @@
+fig, ax = plt.subplots(subplot_kw={"projection": "3d"}, figsize=(7,7))
+
+surf1 = ax.plot_surface(XX, YY, Z.T[0].T, cmap=cm.coolwarm, alpha=0.8)
+surf2 = ax.plot_surface(XX, YY, Z.T[1].T, cmap=cm.coolwarm, alpha=0.8)
```
In this plot I visualised the softmax-function in two dimensions.

The color is for visual ease, but as it is the z-axis in the plot of the softmax-function, it signifies the probability of one of the two events. The z-axis direction is irrelevant, since it is symmetric over the xy-plane in this case.

colinhex commented:
Hi! I addressed the majority of the comments and added a few additional things. E.g. a cross-entropy figure.

colinhex commented on logistic-regression.md:
```diff
@@ -0,0 +1,927 @@
+
+ax.view_init(-18, -172)
+
+plt.show()
```
For the vector plot I adhered to the original script while making it more readable. The content of the vectors have no significance here as long as the geometry is clear and were overlapping in the original graphic. Instead I added a legend with algebraic terms.

colinhex commented on logistic-regression.md:
```diff
@@ -0,0 +1,927 @@
+normal distribution $F_X$ we have:
+
+$$
+    F_{X + c}(x) = p(X + c \leq x) = p(X \leq x - c) = \int_{-\infty}^{x-c}\frac{1}{\sigma\sqrt{2\pi}}e^{-\frac{1}{2}\left(\frac{(t-\mu)^2}{\sigma}\right)}dt
```
Although it is technically a recap at this point and this should be scribed in a previous chapter, I added an explanation for the shift of the mean with the PDF of the normal distribution.

colinhex commented on logistic-regression.md:
```diff
@@ -0,0 +1,927 @@
+$$
+
+Let us visualize this for two normal distributions $\gamma(x)\sim\mathcal{N}(s, 1)$ and $\delta(x)\sim\mathcal{N}(0, 1)$. We
+will shift the mean $s$ of $\gamma(x)$ and calculate the cross entropy along the way!
```
I added a visualisation of the cross-entropy of two normally distributed random variables, when the mean of one is shifted.

vinith2 replied:
The plot looks good. I think you computed the entropy numerically, For gaussians you can compute it analytically as its just a function of its second order moment. If possible can you add this curve as well?

colinhex commented on logistic-regression.md:
```diff
@@ -0,0 +1,927 @@
+
+The term $\sum_{c=1}^C y_{ic} ln(\mu_{ic})$ looks suspiciously similar to **cross-entropy**! This means that
+maximizing the likelihood $\ell(w)$ is equivalent to minimizing the cross-entropy, or to spare the headache *minimizing the negative loss-likelihood.* 
+Why is that the case? We are essentially comparing two probability distributions. On the one hand we have the actual distribution of our data, and on
```
And a explanation that fits the visualisation of the cross-entropy is found in the same section.

vinith2 replied:
regarding the  statement " suspiciously similar ": The equation is an empirical version of the cross entropy. That is for mean E(x), we write it as  $E(X) = \frac{ \sum x_i}{N}$. Similarly the loss is an empirical version of    cross entropy. Something of the form $E_{x \sim \gamma}(\ln(\delta(x)))$. But i think this is ok for now.

colinhex commented:
I addressed your comment regarding gradient descend wrt. gradient steepest descend. However I am finding sources defining terms differently, (gradient descend vs steepest descend vs gradient steepest descend). So [the last commit](https://github.com/swing-research/prbook/pull/6/commits/a4d8c1d78f7fc258d647cb459b2c0c81a4be9ef6) can be taken with a small grain of salt.

vinith2 commented on logistic-regression.md:
```diff
@@ -0,0 +1,942 @@
+First let us consider it with labels $y'_i \in \{0, 1\}$ then for $y'_i = 0$ we have:
+
+$$
+    = -\sum_{i = 1}^k ln\left(\frac{1}{1 + e^{w^Tx_i}}\right)  = \sum_{i = 1}^k ln\left(1 + e^{w^Tx_i}\right)
```
Here you are splitting the sum right? May be add this condition in the sum something like $\sum_{i=0, y_i^{'} = 1}$

vinith2 reviewed (commented):
Hey,
The changes looks good. I have added some comments, but its optional and maybe too much  into the weeds. For now I will keep the PR open.

vinith2 reviewed (approved):
I will approve your changes for now, I will add your name in the headline (for now) so we can evaluate the submission easily.  However there are some minor changes, which we will do it later.
